### DISCUSSION TITLE
Ren'py style TTS support

### DISCUSSION THREAD

**Discussion by zirize** [exult/exult#599] [General]
The Renpy engine implemented tts in a very primitive way, like passing a dialog as an argument and running an external program in the background, but it's very easy to implement than you think.
On Linux it's 'espeak [some dialog]', on MS Windows it's 'say.vbs [some argument] [some dialog]' in the renpy lib.
system(“some_bash_wrapper blah blah &”); I think something like this would work.

With the development of AI, TTS with a natural voice like piper is becoming more accessible, so it would be nice to support this feature in games.

[Example Video](https://youtu.be/vF8N0QbWMLE)

[patch_exult_tts.diff.zip](https://github.com/user-attachments/files/18047704/patch_exult_tts.diff.zip)
[espeak_exult.zip](https://github.com/user-attachments/files/18047727/espeak_exult.zip)

[self-voicing with natural voices with Piper-TTS on Linux](https://www.reddit.com/r/RenPy/comments/1ganueb/selfvoicing_with_natural_voices_with_pipertts_on/)

This patch is a proof of concept and a starting point, but it is far too immature to be used in a real game. I am not a professional C++ programmer, so this is my limitation.

**Comment by zirize** [comment]
[exult_tts.zip](https://github.com/user-attachments/files/18053304/exult_tts.zip)

I've improved the code a bit, but it's still a prototype.

1. the location of tts_wrapper is DATA.
2. supporting book and scroll
3. added a routine to filter out characters that tts can't handle well.

You should modify  tts_wrapper to suit your own TTS.

**Comment by thecommissar** [comment]
This is pretty cool.
